aten/src/ATen/Declarations.cwrap

```diff
@@ -2391,7 +2391,7 @@
     if (self_->dim() == 0) {
       throw std::runtime_error("Input must be 1-d or 2-d");
     }
-    ${THTensor}_diag(${state,}result_->tensor, self_->tensor, diagonal);
+    ${THTensor}_diag(${state,}(THTensor*)(result_->unsafeGetTH(false)), (THTensor*)(self_->unsafeGetTH(false)), diagonal);
```

Comment:
These casts kind of make things worse, because once you set things up so that `THTensor = at::TensorImpl`, you now have a big pile of no-op casts, but if they ever become *not* no-ops you won't notice (because they're C-style unsafe casts.)

Reply:
A safer way to do this is by adding a different, templated function, like `result_->unsafeGetTH<THTensor>`. This will do a `static_cast` in the current impl, but then later it will just static assert that the template argument is `T`, and then at some later point these can all be code-modded to be named a different function.